mlir/lib/Dialect/Linalg/Transforms/HoistVectorTransfers.cpp

```diff
@@ -0,0 +1,267 @@
+    rewriter.setInsertionPoint(reductionForOp);
+    auto *cloneVectorReadOp = rewriter.clone(*vectorReadOpAcc);
+
+    // Code to re-create the reduction and k loop with iter args
+    auto vectorReadOpValue = cloneVectorReadOp->getResult(0);
+    auto newReductionForOp = rewriter.create<scf::ForOp>(
+        reductionForOp.getLoc(), reductionForOp.getLowerBound(),
+        reductionForOp.getUpperBound(), reductionForOp.getStep(),
+        ValueRange{vectorReadOpValue},
+        [&](OpBuilder &rewriterNewReductionForOp, Location locNewReductionForOp,
+            Value ivNewReductionForOp, ValueRange iterArgsNewReductionForOp) {
+          auto newKForOp = rewriter.create<scf::ForOp>(
+              kForOp.getLoc(), kForOp.getLowerBound(), kForOp.getUpperBound(),
+              kForOp.getStep(), iterArgsNewReductionForOp,
+              [&](OpBuilder &rewriterNewKForOp, Location locNewKForOp,
+                  Value ivNewKForOp, ValueRange iterArgsNewKForOp) {
+                IRMapping mapper;
+                mapper.map(reductionForOp.getInductionVar(),
+                           ivNewReductionForOp);
+                mapper.map(kForOp.getInductionVar(), ivNewKForOp);
+
+                for (auto &op : kForOp.getBody()->without_terminator()) {
+                  rewriterNewKForOp.clone(op, mapper);
+                }
+                rewriterNewKForOp.create<scf::YieldOp>(locNewKForOp,
+                                                       iterArgsNewKForOp);
+              });
+          rewriterNewReductionForOp.create<scf::YieldOp>(
+              locNewReductionForOp, newKForOp.getResult(0));
+        });
+
+    // Code to hoist vector transfer write after reduction loop and also to
+    // update the yield of k loop
+    auto newKForOp =
+        llvm::dyn_cast<scf::ForOp>(newReductionForOp.getBody()->front());
+    Value newcontractOpValue;
+    vector::TransferWriteOp vectorWriteOperation;
+    Block *bodyBlock = newKForOp.getBody();
+    for (auto &op : bodyBlock->getOperations()) {
+      if (auto vectorContractOp = llvm::dyn_cast<vector::ContractionOp>(op)) {
+        vectorContractOp.setOperand(vectorContractOp.getNumOperands() - 1,
+                                    newKForOp.getRegionIterArgs()[0]);
+        newcontractOpValue = vectorContractOp.getResult();
+      }
+      if (auto yieldOp = llvm::dyn_cast<scf::YieldOp>(op)) {
+        yieldOp.setOperand(0, newcontractOpValue);
+      }
+      if (auto vectorWriteOp = llvm::dyn_cast<vector::TransferWriteOp>(op)) {
+        vectorWriteOperation = vectorWriteOp;
+      }
+    }
+
+    vectorWriteOperation.setOperand(0, newReductionForOp.getResult(0));
+    vectorWriteOperation->moveBefore(reductionForOp);
+
+    // Erase the old vector contract operation
+    for (auto result : contractOp->getResults()) {
+      for (auto *userOp : result.getUsers()) {
+        userOp->erase();
+      }
+    }
+    contractOp.erase();
```

Comment:
Try to make use of the `replaceAllUsesWith` API. Alternatively, modify the relevant operand once the region containing the contract has been transplanted to the loops with the new iter_args (though this approach is not always safe so probably best avoided as well).